core/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala

```diff
@@ -130,20 +131,22 @@ case class UpdateCommand(
       val newTarget = DeltaTableUtils.replaceFileIndex(target, fileIndex)
       val data = Dataset.ofRows(sparkSession, newTarget)
       val updatedRowCount = metrics("numUpdatedRows")
-      val updatedRowUdf = DeltaUDF.boolean { () =>
-        updatedRowCount += 1
-        true
-      }.asNondeterministic()
-      val pathsToRewrite =
+
+      val (pathsToRewrite, observedMetrics) =
         withStatusCode("DELTA", UpdateCommand.FINDING_TOUCHED_FILES_MSG) {
-          data.filter(new Column(updateCondition))
-            .select(input_file_name())
-            .filter(updatedRowUdf())
```

Comment:
Do you know why Spark doesn't do the column pruning properly? Although we can change the code to work around it, it's better to also fix it in Spark side as well.

Reply:
I think it's because `PhysicalOperation` is used to determine nested fields to prune, and non-deterministic `Project`'s aren't included in `PhysicalOperation`'s. I feel like there's a lot of optimizations that just don't apply to non-deterministic functions, so if there's ways to avoid using them it's probably good, especially when it's just to collect metrics.

Reply:
Or in this case, non-deterministic filter expressions don't get included in `PhysicalOperation`'s. I'm not exactly sure when/how the empty `Project` gets added in the plan in the description. Interestingly nested fields don't get pushed through a `CollectMetrics` operation, which prevents full schema pruning with this method for merging, but I already have a working Spark patch to enable that for the next Spark release if we go down the CollectMetrics vs non-determinstic UDF approach

Reply:
And actually it's not just the metrics, but also the `input_file_name` that I thing combined or one of the other causes the problem, it's hard to say exactly, but potentially use metadata fields instead of `input_file_name` in the future too to help address these weird cases caused by non-deterministic functions

Reply:
> I think it's because PhysicalOperation is used to determine nested fields to prune, and non-deterministic Project's aren't included in PhysicalOperation's. 

IIRC, the non-deterministic UDF makes Spark not be able to know which columns are needed? Technically, Spark can still do the nested column pruning on the non-deterministic UDF if the UDF doesn't read these columns. Right? Do you know the exact optimization rule?

Trying to understand the entire problem better so that we can apply the same approach to other places that are using  non-deterministic UDFs.

Reply:
Yeah still trying to figure out the exact reason/code path. But I already have a fix for `Delete` which has the same problem, and a partial fix for `Merge` (which has additional issues that might only be fixable come Spark 3.4), which are the only other places it looks like the non-deterministic func for metrics thing is used.

Reply:
Ok so while digging into the why, I was testing out on Spark master and couldn't recreate the issue. After a lot of digging, I'm pretty sure https://github.com/apache/spark/pull/37176 will fix this issue in Spark 3.4+. Gonna try to test all the cases (update/delete/merge) with latest Spark master and see if it addresses all the current limitations